metricbeat/module/system/raid/blockinfo/parser.go

```diff
@@ -0,0 +1,207 @@
+	}
+
+	current, err := strconv.ParseInt(matches[0], 10, 64)
+	if err != nil {
+		return SyncStatus{}, errors.Wrap(err, "could not parse data sync_completed")
+	}
+
+	total, err := strconv.ParseInt(matches[1], 10, 64)
+	if err != nil {
+		return SyncStatus{}, errors.Wrap(err, "could not parse data sync_completed")
+	}
+
+	return SyncStatus{Complete: current, Total: total}, nil
+
+}
+
+//Create a new disk object, parsing any needed fields
+func newMD(path string) (MDDevice, error) {
+	var dev MDDevice
+
+	dev.Name = filepath.Base(path)
+	size, err := parseIntVal(filepath.Join(path, "size"))
+	if err != nil {
+		return dev, errors.Wrap(err, "could not get device size")
+	}
+	dev.Size = size
+
+	//This is the count of 'active' disks
+	// active, err := parseIntVal(filepath.Join(path, "md", "raid_disks"))
+	// if err != nil {
+	// 	return dev, errors.Wrap(err, "could not get raid_disks")
+	// }
+	// dev.ActiveDisks = active
+
+	//RAID array state
+	state, err := ioutil.ReadFile(filepath.Join(path, "md", "array_state"))
+	if err != nil {
+		return dev, errors.Wrap(err, "could not open array_state")
+	}
+	dev.ArrayState = strings.TrimSpace(string(state))
+
+	//get total disks
+	disks, err := getDisks(path)
+	if err != nil {
+		return dev, errors.Wrap(err, "could not get disk data")
+	}
+	dev.DiskStates = disks
+
+	//sync action and sync status will only exist for redundant raid levels
+	redundant, err := isRedundant(path)
+	if err != nil {
+		return dev, errors.Wrap(err, "error getting raid level")
+	}
+	if redundant {
+
+		//Get the sync action
+		//Will be idle if nothing is going on
+		syncAction, err := ioutil.ReadFile(filepath.Join(path, "md", "sync_action"))
+		if err != nil {
+			return dev, errors.Wrap(err, "could not open sync_action")
+		}
+		dev.SyncAction = strings.TrimSpace(string(syncAction))
+
+		//sync status
+		syncStats, err := getSyncStatus(path, dev.Size)
+		if err != nil {
+			return dev, errors.Wrap(err, "error getting sync data")
+		}
+
+		dev.SyncStatus = syncStats
+	}
+
+	return dev, nil
+}
+
+//get all the disks associated with an MD device
+func getDisks(path string) (DiskStates, error) {
+	//so far, haven't found a less hacky way to do this.
+	devices, err := filepath.Glob(filepath.Join(path, "md", "dev-*"))
+	if err != nil {
+		return DiskStates{}, errors.Wrap(err, "could not get device list")
+	}
+
+	var disks DiskStates
+	disks.States = common.MapStr{}
+	//This is meant to provide a 'common status' for disks in the array
+	//see https://www.kernel.org/doc/html/v4.15/admin-guide/md.html#md-devices-in-sysfs
+	for _, disk := range devices {
+		disk, err := getDisk(disk)
+		if err != nil {
+			return DiskStates{}, err
+		}
+
+		switch disk {
+		case "faulty", "blocked", "write_error", "want_replacement":
+			disks.Failed++
+		case "in_sync", "writemostly", "replacement":
+			disks.Active++
+		case "spare":
+			disks.Spare++
+		default:
+			disks.Unknown++
```

Comment:
Just log the state found at the debug level, so we can investigate and add it to some of the other categories.